query/services/ingestion.py

```diff
@@ -0,0 +1,199 @@
+from query.tables.loaded_tag import append_loaded_tags
+from query.tables.product import (
+    create_minimal_product,
+    delete_products,
+    get_minimal_product,
+    product_fields,
+    update_products_from_staging,
+)
+from query.tables.product_country import fixup_product_countries
+from query.tables.product_ingredient import create_ingredients_from_staging
+from query.tables.product_tags import TAG_TABLES, create_tags_from_staging
+from query.tables.settings import get_last_updated, set_last_updated
+
+logger = logging.getLogger(__name__)
+min_datetime = datetime(1, 1, 1, tzinfo=timezone.utc)
+batch_size = 10000
+
+
+async def get_process_id(connection):
+    return await connection.fetchval("SELECT pg_current_xact_id()")
+
+
+def int_or_none(value):
+    return None if value == None else int(value)
+
+
+async def import_with_filter(filter: Dict, source: Source) -> datetime:
+    max_last_updated = min_datetime
+    found_product_codes = []
+    codes_specified = "code" in filter and "$in" in filter["code"]
+
+    async with database_connection() as connection:
+        await connection.execute(
+            "CREATE TEMP TABLE product_temp (id int PRIMARY KEY, last_updated timestamptz, data jsonb)"
+        )
+        try:
+            process_id = await get_process_id(connection)
+            projection = {
+                key: True
+                for key in (list(TAG_TABLES.keys()) + list(product_fields.keys()))
+            }
+            for obsolete in [False, True]:
+                update_count = 0
+                skip_count = 0
+                async with find_products(filter, projection, obsolete) as cursor:
+                    async for product_data in cursor:
+                        # Fall back to last_modified_t if last_updated_t is not available
+                        try:
+                            last_updated = datetime.fromtimestamp(
+                                product_data.get(
+                                    "last_updated_t",
+                                    product_data.get("last_modified_t"),
+                                ),
+                                timezone.utc,
+                            )
+                        except TypeError:
+                            logger.warning(
+                                f"Product: {product_data['code']}. Invalid last_updated_t: {product_data.get('last_updated_t')}, or last_modified_t: {product_data.get('last_modified_t')}."
+                            )
+                            last_updated = min_datetime
+                        max_last_updated = max(last_updated, max_last_updated)
+
+                        product_code = product_data["code"]
+                        if codes_specified:
+                            found_product_codes.append(product_code)
+
+                        existing_product = await get_minimal_product(
+                            connection, product_code
+                        )
```

Comment:
Is this efficient to do this one by one ? It's easier to read though, but maybe, first getting product data, then fetching existing and then processing would still be readable ?

Reply:
I considered refactoring this but it gets complicated as we don't want to insert the full product JSON into the temporary table if we are going to skip it anyway, so that could make the logic quite complicated. It also doesn't seem to be a big part of where the time is being spent